[DO NOT MERGE] Add blind version signing for a generic request

Aerilym opened the pull request:
Should be merged to the new repo instead: https://github.com/session-foundation/libsession-util/pull/4
this one should be closed

Aerilym committed:
Add blind version signing for a generic request

tewinget commented on src/blinding.cpp:
```diff
@@ -456,6 +456,22 @@ ustring blind15_sign(ustring_view ed25519_sk, std::string_view server_pk_in, ust
     return result;
 }
 
+ustring blind_version_sign_request(ustring_view ed25519_sk, uint64_t timestamp, ustring_view method, ustring_view path, std::optional<ustring_view> body) {
+    auto [pk, sk] = blind_version_key_pair(ed25519_sk);
+
+    // Signature should be on `TIMESTAMP || METHOD || PATH || BODY_HASH`
+    ustring buf;
+    buf.reserve(10 + 6 + path.size() + body.value_or({}).size());
+    buf += to_unsigned_sv(std::to_string(timestamp));
+    buf += to_unsigned_sv(method);
+    buf += to_unsigned_sv(path);
+
+    if (body)
+        buf += blake2b(*body, 64);
+
+    return ed25519::sign({sk.data(), sk.size()}, buf);
+}
```
```suggestion
ustring blind_version_sign_request(ustring_view ed25519_sk, uint64_t timestamp, ustring_view method, ustring_view path, std::optional<ustring_view> body) {
    auto [pk, sk] = blind_version_key_pair(ed25519_sk);

    // Signature should be on `TIMESTAMP || METHOD || PATH || BODY`
    ustring buf;
    buf.reserve(10 + 6 + path.size() + (body ? body->size() : 0));
    buf += to_unsigned_sv(std::to_string(timestamp));
    buf += to_unsigned_sv(method);
    buf += to_unsigned_sv(path);

    if (body)
        buf += *body;

    return ed25519::sign({sk.data(), sk.size()}, buf);
}
```

tewinget replied:
`ed25519::sign` is already doing a hash with the message (`buf` here) as part of the hashed data, so I don't think hashing before appending to the buffer is necessary.

tewinget replied:
Github won't let me make suggestions outside the default-visible context here, but if desired the `blind_version_sign` function below could be rewritten to use this function (passing `std::nullopt` as body)

Aerilym committed:
Change body hash to body

Co-authored-by: Thomas Winget <tewinget@gmail.com>

Aerilym committed:
Create headers for blind version signing

tewinget committed:
blind_version_sign_request C API and CPP test

Bilb commented on src/blinding.cpp:
```diff
@@ -456,6 +456,27 @@ ustring blind15_sign(ustring_view ed25519_sk, std::string_view server_pk_in, ust
+
+    // Signature should be on `TIMESTAMP || METHOD || PATH || BODY`
+    ustring buf;
+    buf.reserve(10 + 6 + path.size() + (body ? body->size() : 0));
```
should include method.size() too right?

tewinget replied:
Could do.  Honestly I was tempted to remove the `reserve` altogether as it's not going to make a meaningful difference.  But yes, it probably should actually compute the correct size if it's going to reserve.  Can amend, no problem.

tewinget replied:
Done.  Also had forgotten to run `clang-format`, so did that too.

Bilb commented on include/session/blinding.hpp:
```diff
@@ -168,6 +169,19 @@ ustring blind15_sign(ustring_view ed25519_sk, std::string_view server_pk_in, ust
+ustring blind_version_sign_request(
+        ustring_view ed25519_sk,
+        uint64_t timestamp,
+        ustring_view method,
+        ustring_view path,
```
should those be string_view, or we don't care?

tewinget replied:
Good question, and I'm not sure.  I don't know exactly the use case/scenario, but practically speaking they can be whichever is more convenient.

jagerman replied:
`std::string_view`.  Neither of them make any sense containing binary data.

jagerman replied:
(And string_view is always more convenient.  Plus there's the clang+libc++ issues with ustring_views that might necessitate replacing them with spans, which isn't going to be a fun conversion).

tewinget replied:
Changed to `string_view`

tewinget committed:
more correct string reserve, also lint

RyanRory commented on src/blinding.cpp:
```diff
@@ -583,6 +604,33 @@ LIBSESSION_C_API bool session_blind25_sign(
+        const unsigned char* ed25519_seckey,
+        size_t timestamp,
+        const unsigned char* method,
+        size_t method_len,
```
Seems like we don't need the length of method, path, and body, can we clean them up?

tewinget replied:
We could remove them, but there are a couple reasons I think it's better to have them.  Without the size parameter, the string must be null-terminated (which is likely to be the case, but I think there's no harm in guarding against it not being the case) and most string-y containers in most languages already know their length, so computing it here (the constructor for the string view would do so) is redundant.

My experience is that C APIs which take an arbitrary length of data tend to take a pointer to the beginning and a length, but if it is decided that having the length parameters is cumbersome I'm more than willing to change it.

jagerman replied:
In general, for things like method and path (which are text, i.e. non-binary, strings) we've stuck with using `const char*` (and no size) in the C API wrappers, with a `unsigned char*` used specifically for known-fixed size binary data (such as seckey) and `unsigned char*, size_t` pair for variable-sized binary data (such as body).  Null-terminated strings are generally extremely easy to wrap (more so than an `unsigned char*, size_t` pair), and so this also signals to the API wrapper that that isn't binary, but is rather a plain text string value.

tewinget replied:
Amended method and path to use `string_view`, leaving body as `ustring_view` as per Jason's recommendation.  C API mirrors that change.

jagerman commented on include/session/blinding.h:
```diff
@@ -113,6 +113,22 @@ LIBSESSION_EXPORT bool session_blind25_sign(
+/// Returns a version-blinded signature.
+LIBSESSION_EXPORT bool session_blind_version_sign_request(
+        const unsigned char* ed25519_seckey, /* 64 bytes */
+        size_t timestamp,
```
A timestamp value should ideally be a 64-bit type; probably an `int64_t` or `uint64_t`.  (`size_t` is only 32-bits on 32-bit platforms like ARMv7 and x86).

tewinget replied:
Applies to `session_blind_version_sign` as well, fixed both.

tewinget committed:
use string_view instead of ustring_view where reasonable

tewinget committed:
size_t -> uint64_t in C API in a couple places (timestamp)

jagerman commented on src/blinding.cpp:
```diff
@@ -456,6 +456,27 @@ ustring blind15_sign(ustring_view ed25519_sk, std::string_view server_pk_in, ust
+    if (body)
+        buf += *body;
+
+    return ed25519::sign({sk.data(), sk.size()}, buf);
```
This call to `sign` seemed quite wrong to me, because the comment of that `sign` function reads:

```C++
/// Takes the Ed25519 secret key (64 bytes, or 32-byte seed).  Returns the blinded public key and
/// private key (NOT a seed).
```
which would make this a bit undesirable.  But, it turns out, the *comment* is wrong, and what that function actually does is generate a new seed by cryptographic hash of the current seed.  And so it's okay.

BUT: we should amend that comment, to read: "Returns the blinded public key and blinded libsodium seed value."

jagerman reviewed (approved):
One requested change to a very wrong function comment that this PR uses, but other than that this looks good to me.

Bilb committed:
chore: fix comment for blind_version_key_pair returning a seed

jagerman committed:
Make windows-x86 static "allow fail"

It will take more investigation to coax this build into working, and
this is not an important release target.

Bilb commented:
merged on the session-foundation repo
